harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,352 @@
+        DeepSpeedTrial.
+        """
+        optimizations_config = self.env.experiment_config.get_optimizations_config()
+        if not optimizations_config.get("average_training_metrics", False):
+            logging.warning(
+                "DeepSpeedTrial always averages training metrics across data parallel ranks."
+            )
+        self._average_training_metrics = True
+        mixed_precision_val = optimizations_config.get("mixed_precision", "O0")
+        if mixed_precision_val != "O0":
+            raise det.errors.InvalidExperimentException(
+                "Mixed precision is specified through the deepspeed config instead of the "
+                "Determined experiment config.",
+            )
+        aggregation_frequency = optimizations_config.get("aggregation_frequency", 1)
+        if aggregation_frequency > 1:
+            raise det.errors.InvalidExperimentException(
+                "Gradient aggregation is specified through the deepspeed config instead of the "
+                "Determined experiment config.",
+            )
+
+    def get_global_batch_size(self) -> int:
+        return self._global_batch_size
+
+    def get_per_slot_batch_size(self) -> int:
+        return self._per_slot_batch_size
+
+    def wrap_mpu(self, mpu: det_ds.ModelParallelUnit) -> det_ds.ModelParallelUnit:
+        if self._called_wrap_mpu:
+            raise det.errors.InvalidExperimentException(
+                "Only one MPU can be passed to DeepSpeedTrialContext.  "
+                "Please make sure wrap_mpu is only called once in the trial definition."
+            )
+        self._called_wrap_mpu = True
+        old_mpu = self._mpu
+        self._mpu = mpu
+        if old_mpu.get_data_parallel_world_size() != mpu.get_data_parallel_world_size():
+            (
+                self._per_slot_batch_size,
+                self._global_batch_size,
+            ) = self._calculate_batch_sizes()
+            logging.warning(
+                "Wrapped MPU uses model parallelism.  Changing per slot batch size "
+                f"to {self.get_per_slot_batch_size()} and "
+                f"global batch size to {self.get_global_batch_size()}"
+            )
+        return mpu
+
+    @property
+    def mpu(self) -> det_ds.ModelParallelUnit:
+        return self._mpu
+
+    def wrap_model_engine(self, model: torch.nn.Module) -> torch.nn.Module:
+        """Returns a wrapped model engine.
+
+        In the background, we perform checks to properly handle pipeline parallelism if
+        the model engine is a PipelineEngine.  We also recompute batch sizes to match
+        the deepspeed config.
+        """
+
+        if self.env.managed_training:
+            model = model.to(self.device)
+
+        # Pipeline parallel model engine has its own MPU that we will use here.
+        if isinstance(model, deepspeed.PipelineEngine):
+            self._use_pipeline_parallel = True
+            self._mpu = det_ds.DeepSpeedMPU(model.mpu)
```

Comment:
this looks like it has an ordering problem with `wrap_mpu()`

Reply:
we automatically set mpu if the user passes a `PipelineEngine`.  We do not record this as a call to `wrap_mpu` to allow the user to optionally override the mpu if they want to (e.g. in order to build dataloader on every rank or have every rank report metrics).

Reply:
In order to make it easy to check that only one mpu is active to simplify supporting multiple model_engines, we will make users call `wrap_mpu` if they are using pipeline parallelism.